### PR TITLE
[WIP] Build SDK with contrib repo exporters and examples - `BUILD_CONTRIB` build option

### LINKED ISSUE #713: Allow to build SDK + contrib repo via `BUILD_CONTRIB` build option
**Is your feature request related to a problem?** _Not really a problem..._

OpenTelemetry C++ SDK contrib repo:

https://github.com/open-telemetry/opentelemetry-cpp-contrib

may contain additional "non-standard" exporters that are maintained by community, e.g. `fluentd` exporter, for example.

**Describe the solution you'd like**

I would like to implement a build option for the main repo that allows to build contrib - `BUILD_CONTRIB` :
- by turning on a custom build option; OR
- by specifying the path to local checked out contrib snapshot, OR to alternate "contrib" overlay.

**Describe alternatives you've considered**

We could have added contrib repo as a submodule. Adding it as submodule presents other maintenance challenges.

**Additional context**

- **CMake** : when the SDK is built `BUILD_CONTRIB`, it should include the `CMakeLists.txt` from contrib path using `add_subdirectory(/path/to/contrib)` , that way the entire build tree of contrib repo is also included in the build.

- **Bazel** : contrib may be added as symlink (Unix) or directory juniction (Windows) at `/contrib` path. Bazeil would pick it up.
What it gives is ability to build a distro of SDK with additional exporters.

Note that the `contrib`:
- could be our contrib with community provided OSS exporters (mentioned above); or
- could be just totally different PRIVATE CUSTOM vendor contrib repository, with private closed-source non-standard vendor exporters.

**What we achieve is a single build that builds the SDK with custom additions?**

- Continuous CI of contrib against the main repository.
- Running all tests of contrib plus the main repository.
- Opportunity to customize the build targets and provide your own combined build target.
- Contrib knows the main repo targets; it can cover extra steps, such as custom binary packaging (e.g. nuget)

I'd like to illustrate how that would work for the following example:
- `fluentd` exporter lives in contrib repo.
- build option in the main allows to build with a snapshot of contrib repo overlaid on top of the main SDK repo.

**Bonus points**

Once the change is merged, we could have added a CI that implements an _OPTIONAL_ build check: verify that refactor in the main is not breaking anything significant in contrib. By building the main `BUILD_CONTRIB` against the latest `https://github.com/open-telemetry/opentelemetry-cpp-contrib` (main) branch snapshot of contrib. That way if refactor happens in the main that breaks any of the contrib exporters, we would know right away.

### PR DESCRIPTION
**Pull request by maxgolov**
Fixes #713 

## Changes

New CMake build options that allow to build the main OpenTelemetry C++ SDK with optional community contributions from [opentelemetry-cpp-contrib](https://github.com/open-telemetry/opentelemetry-cpp-contrib) repo.

- first option is enabled by environment variable `${OPENTELEMETRY_CONTRIB_PATH}`. This is, in its spirit, is similar how OpenCV handles its optional `modules`. If this variable is set, then the build adds the directory as subdirectory using `add_subdirectory($ENV{OPENTELEMETRY_CONTRIB_PATH})`. So the contents of this directory must contain `CMakeLists.txt` and other projects within, and get built alongside with the rest of SDK. This first option works with any CMake version.

- second option is enabled by CMake config option called `BUILD_CONTRIB`. If it is enabled, then the build would attempt to fetch `https://github.com/open-telemetry/opentelemetry-cpp-contrib` into source tree. And add it using `add_subdirectory`. Again, the remote repo must contain `CMakeListst.txt`. I'm temporarily adding my remote branch for illustrative purposes. This option _requires_ CMake 3.11+, but since it's optional - hey! if you have an older CMake, then just don't use it. OR upgrade your CMake.

Main motivation is to enable ONE build of SDK + community contributed dependencies, e.g. my immediate need is to build with `fluentd` exporter (_that is a non-standard exporter, not prescribed by mainline spec, so belongs to contrib_).

## Example build

- Step 1: I included an example that works with Visual Studio 2019, that adds a new build combo called `stdlib-x64-Release-contrib. Generate CMake cache with `cmake -DBUILD_CONTRIB=ON`

- Step 2: Run the build :

```console
>------ Build All started: Project: opentelemetry-cpp, Configuration: stdlib-x64-Release-contrib ------
...
[236/239] Linking CXX executable ext\test\zpages\threadsafe_span_data_test.exe
  [237/239] Linking CXX executable ext\test\zpages\tracez_data_aggregator_test.exe
  [238/239] Building CXX object out\vs2019\stdlib-x64-Release-contrib\_deps\contrib-src\exporters\fluentd\CMakeFiles\fluentd_recordable_test.dir\test\fluentd_recordable_test.cc.obj
  [239/239] Linking CXX executable out\vs2019\stdlib-x64-Release-contrib\_deps\contrib-src\exporters\fluentd\fluentd_recordable_test.exe
```

Note that the last executable - `fluentd` exporter test has been added from contrib repo. Now we run all tests for all components, including additional community-contributed components (exporters, examples, etc.) in one single workspace. This enables CI to validate that contrib is building successfully with latest `opentelemetry-cpp` repository.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -210,6 +214,25 @@ if(WITH_EXAMPLES)
...
+  add_subdirectory($ENV{OPENTELEMETRY_CONTRIB_PATH})
+endif()
+
+if(WITH_CONTRIB)
+  # This option requires CMake 3.11+: add standard remote contrib to build tree.
+  include(FetchContent)
+  FetchContent_Declare(
+    contrib
+    GIT_REPOSITORY "https://github.com/open-telemetry/opentelemetry-cpp-contrib"
+    GIT_TAG "maxgolov/fluentd_exporter")
```
Comment (maxgolov) [inline CMakeLists.txt:228]: This is temporary. Will change it to `main` before the merge.

File: CMakeSettings.json [hunk 2]
```diff
@@ -1,126 +1,157 @@
...
+            "buildRoot": "${projectDir}\\out\\vs2019\\${name}",
+            "installRoot": "${projectDir}\\out\\vs2019\\${name}\\install",
+            "cmakeCommandArgs": "",
+            "buildCommandArgs": "",
+            "ctestCommandArgs": "",
+            "cmakeToolchain": "",
+            "inheritEnvironments": [ "msvc_x64_x64" ],
+            "variables": [
+                {
+                    "name": "WITH_CONTRIB",
```
Comment (maxgolov) [inline CMakeSettings.json:130]: This is funny format that Visual Studio 2019 uses for `CMake` build combos. This combo, for illustrative purposes, is built with all possible exporters and `WITH_CONTRIB`.

Comment (maxgolov) [reply CMakeSettings.json:130]: The other second option is to specify the path explicitly:

```json
        {
            // These environment variables apply to all configurations.
            "environments": [
                {
                    "OPENTELEMETRY_CONTRIB_PATH": "${projectDir}\\..\\opentelemetry-cpp-contrib"
                }
            ],
            "name": "stdlib-x64-Release-contrib"
        }
```

Then a local custom snapshot can be used, e.g. my checked out tree of contrib side-by-side with the main is built in one build when I build the main, with all artifacts and tests ran at the same time.

**Comment by lalitb** [comment]
@jsuereth, @pyohannes  - Need your feedback please for this PR - As mentioned in the description, this brings the cmake options `BUILD_CONTRIB` to build the contrib repo as part of the main repo build. The plan is to use this option in the nightly CI build of `contrib` repo to test against the latest `main` repo commits.

**Pull request closed without merging**